File: examples/csharp/src/examples/Example02EnumerateAdapters.cs
using DXBridgeExamples.DxBridge;

namespace DXBridgeExamples.Examples;

internal static class Example02EnumerateAdapters
{
    public static int Run(string[] args)
    {
        var cli = new CommandLine(args, new[] { "--dll", "--backend" }, Array.Empty<string>());
        if (cli.HelpRequested)
        {
            Console.WriteLine("Usage: Example02EnumerateAdapters [--dll PATH] [--backend auto|dx11|dx12]");
            return 0;
        }

        using var bridge = new DXBridge(cli.GetString("--dll"));
        var initialized = false;
        try
        {
            Console.WriteLine($"Loaded: {bridge.DllPath}");
            bridge.Init(DXBridge.BackendFromName(cli.GetString("--backend", "auto")));
            initialized = true;

            var adapters = bridge.EnumerateAdapters();
            if (adapters.Count == 0)
            {
                Console.WriteLine("No adapters reported by DXBridge.");
                return 0;
            }

            Console.WriteLine($"Found {adapters.Count} adapter(s):");
            foreach (var adapter in adapters)
            {
                var kind = adapter.IsSoftware != 0 ? "software" : "hardware";
                Console.WriteLine($"- #{adapter.Index}: {DXBridge.AdapterDescription(adapter)} [{kind}] | VRAM={DXBridge.FormatBytes(adapter.DedicatedVideoMem)} | Shared={DXBridge.FormatBytes(adapter.SharedSystemMem)}");
            }
        }
        finally
        {
            if (initialized)
            {
                bridge.Shutdown();
            }
        }

        return 0;
    }
}

File: examples/java/src/examples/Example02EnumerateAdapters.java
package examples;

import java.util.List;

import dxbridge.CommandLine;
import dxbridge.DXBridge;
import dxbridge.DXBridgeLibrary;

public final class Example02EnumerateAdapters {
    public static void main(String[] args) {
        CommandLine cli = parse(args);
        if (cli.isHelpRequested()) {
            printUsage();
            return;
        }

        DXBridge bridge = new DXBridge(cli.getString("--dll", null));
        boolean initialized = false;

        try {
            System.out.println("Loaded: " + bridge.getDllPath().getAbsolutePath());
            bridge.init(DXBridge.backendFromName(cli.getString("--backend", "auto")));
            initialized = true;

            List<DXBridgeLibrary.DXBAdapterInfo> adapters = bridge.enumerateAdapters();
            if (adapters.isEmpty()) {
                System.out.println("No adapters reported by DXBridge.");
                return;
            }

            System.out.println("Found " + adapters.size() + " adapter(s):");
            for (DXBridgeLibrary.DXBAdapterInfo adapter : adapters) {
                String kind = adapter.is_software != 0 ? "software" : "hardware";
                System.out.println(
                    "- #" + adapter.index + ": " + DXBridge.adapterDescription(adapter)
                    + " [" + kind + "] | VRAM=" + DXBridge.formatBytes(adapter.dedicated_video_mem)
                    + " | Shared=" + DXBridge.formatBytes(adapter.shared_system_mem)
                );
            }
        } finally {
            if (initialized) {
                bridge.shutdown();
            }
        }
    }

    private static CommandLine parse(String[] args) {
        return new CommandLine(args, CommandLine.options("--dll", "--backend"), CommandLine.options());
    }

    private static void printUsage() {
        System.out.println("Usage: Example02EnumerateAdapters [--dll PATH] [--backend auto|dx11|dx12]");
    }
}
